### PR TITLE
define multiple overmap terrains in a single json object

### PR DESCRIPTION
**Pull request by vickio**
From the included `MAPGEN.md` documentation:

> As an alternative to "om_terrain", "om_terrain_multi" allows for multiple overmap terrains to be defined using a single json object, with the "rows" property expanding in blocks of 24x24 characters to accommodate as many overmap terrains as are listed here. The terrain ids are specified using a nested array of strings which represent the rows and columns of overmap terrain ids (found in overmap_terrain.json) that are associated with the "rows" property described in section 2.1 of this document. To use "om_terrain_multi", "om_terrain" must not be specified.
>
> Characters mapped using the "terrain", "furniture", or any of the special mappings ("items", "monsters", etc) will be applied universally to all of the listed overmap terrains.
>
> Placing things using x/y coordinates ("place_monsters", "place_loot", "add", etc) works using the full extended coordinates beyond 24x24. An important limitation is that ranged random coordinates (such as "x": [ 10, 18 ]) must not cross the 24x24 terrain boundaries. Ranges such as [ 0, 23 ] and [ 50, 70 ] are valid, but [ 0, 47 ] and [ 15, 35 ] are not because they extend beyond a single 24x24 block.
>
> Example:
```
"om_terrain_multi": [
  [ "apartments_mod_tower_NW", "apartments_mod_tower_NE" ],
  [ "apartments_mod_tower_SW", "apartments_mod_tower_SE" ]
]
```
>
> In this example, the "rows" property should be 48x48, with each quadrant of 24x24 being associated with each of the four apartments_mod_tower overmap terrain ids specified.

In this PR I have also included a straightforward conversion of the recently merged `apartment_mod.json` to show the major readability advantage of this method. It can be used like that to define all the om terrains for an entire om special in a single object. Or since mapgen doesn't actually understand any relationship between the om terrains, it could be used to share `"terrain"`, `"furniture"` and other mappings between several otherwise unrelated om terrains, or between different variations of the same om terrain, simply by sticking them together in the same `"rows"` property.

### CODE REVIEW

File: src/mapgen.cpp [hunk 1]
```diff
@@ -361,8 +362,25 @@ void load_mapgen( JsonObject &jo )
         }
     } else if ( jo.has_string( "om_terrain" ) ) {
         load_mapgen_function(jo, jo.get_string( "om_terrain" ), -1);
+    } else if( jo.has_array( "om_terrain_multi" ) ) {
```
Comment (codemime) [inline src/mapgen.cpp]: I suppose you don't need another name. `om_terrain` can be treated either as string or an array of strings (depending on its content). This is also more robust: `om_terrain_multi` won't be silently ignored in favor of `om_terrain`.

File: src/mapgen.cpp [hunk 2]
```diff
@@ -361,8 +362,25 @@ void load_mapgen( JsonObject &jo )
...
+        JsonArray rows = jo.get_array( "om_terrain_multi" );
+        int x_offset = 0;
+        int y_offset = 0;
+        while( rows.has_more() ) {
+            JsonArray row_items = rows.next_array();
+            while( row_items.has_more() ) {
+                std::string mapgenid = row_items.next_string();
```
Comment (codemime) [inline src/mapgen.cpp]: Can be `const`.

File: src/mapgen.cpp [hunk 3]
```diff
@@ -361,8 +362,25 @@ void load_mapgen( JsonObject &jo )
...
+                mapgen_function *mgfunc = load_mapgen_function( jo, mapgenid, -1, x_offset, y_offset );
+                if( mgfunc != NULL ) {
```
Comment (codemime) [inline src/mapgen.cpp]: Please use `nullptr` instead.

File: src/mapgen.cpp [hunk 4]
```diff
@@ -407,16 +425,18 @@ bool mapgen_function_json::check_inbounds( const jmapgen_int & var ) const {
     return true;
 }
 
-mapgen_function_json::mapgen_function_json( std::string s, int const w )
+mapgen_function_json::mapgen_function_json( std::string s, const int w, const int x_grid_offset, const int y_grid_offset )
```
Comment (codemime) [inline src/mapgen.cpp]: Since you've touched this line: `const std::string &s`.

File: src/mapgen.cpp [hunk 5]
```diff
@@ -1101,6 +1128,32 @@ class jmapgen_make_rubble : public jmapgen_piece {
...
 
+jmapgen_objects::jmapgen_objects( const int x_offset, const int y_offset, const int mapgensize )
+: x_offset( x_offset )
+, y_offset( y_offset )
+, mapgensize( mapgensize )
+{}
+
+bool jmapgen_objects::check_bounds( const jmapgen_place place, JsonObject &jso )
+{
+    if( place.x.val < 0 || (size_t)place.x.val > mapgensize - 1 ||
```
Comment (codemime) [inline src/mapgen.cpp]: This comparison is potentially dangerous: `(size_t)place.x.val > mapgensize - 1`. If `mapgensize` is zero (unlikely, but anyways), the condition is always `true` due to unsigned overflow.

Comment (AMurkin) [inline src/mapgen.cpp]: @codemime maybe always `false`?

Comment (codemime) [inline src/mapgen.cpp]: Yeah, sure thing. Thanks.

**Comment by vickio** [comment]
@codemime thanks for the review. I think I've addressed the issues you noted.

**Comment by codemime** [comment]
Has a trivial merge conflict. Otherwise it's good to go.

**Comment by vickio** [comment]
Rebased to master and resolved the conflict.

**Comment by vickio** [comment]
The conflicting commit got reverted after I resolved it, so I rebased again and re-resolved it.

**Pull request merged**